File: common/modules/debrid/realdebrid.js
// relative import keeps this module loadable under plain Node for API tests
import DebridService, { DebridError, DebridAuthError, DebridNotCachedError, DebridTimeoutError, DebridUnavailableError, archiveRx } from './service.js'
import { Availability } from './availability.js'
import Debug from 'debug'
const debug = Debug('ui:debrid')

const API = 'https://api.real-debrid.com/rest/1.0'
const LIST_LIMIT = 1_000 // the whole account in one request, newest first

/**
 * What each torrent status means for playback. Magnet conversion and file selection are left out
 * on purpose: they are moments in a torrent's life, not outcomes, so they answer nothing.
 * @type {Record<string, string>}
 */
const statusAvailability = {
  downloaded: Availability.CACHED,
  // being fetched fresh, so Real-Debrid can serve it eventually but not now
  queued: Availability.AVAILABLE,
  downloading: Availability.AVAILABLE,
  uploading: Availability.AVAILABLE,
  compressing: Availability.AVAILABLE,
  // will never complete
  magnet_error: Availability.UNAVAILABLE,
  error: Availability.UNAVAILABLE,
  virus: Availability.UNAVAILABLE,
  dead: Availability.UNAVAILABLE
}

// error_code values worth explaining, anything else falls back to the API's own message
const errorMessages = {
  8: 'Invalid Real-Debrid API key',
  9: 'Real-Debrid denied the request, check the account permissions',
  21: 'Too many active Real-Debrid downloads, wait for one to finish',
  23: 'This Real-Debrid account has exhausted its traffic',
  34: 'Real-Debrid is rate limiting this account, try again shortly',
  35: 'Real-Debrid will not serve this file, pick a different release',
  36: 'Real-Debrid fair usage limit reached'
}
// only these mean the key or account is the problem, the rest are per-request
const authCodes = [8, 9]
// the account cannot take more work right now: rate limited, or at its active torrent cap
const throttleCodes = [21, 34]

/**
 * Real-Debrid implementation, see https://api.real-debrid.com/
 *
 * Two API quirks shape this client:
 * - `/torrents/instantAvailability` is disabled (403 `disabled_endpoint`), so a release can only
 *   be asked about by adding the magnet and reading the status back. Hence `probeAvailability`,
 *   and the base class cap on how many of those a search may cost.
 * - Selecting multiple files can serve one RAR archive instead of individual links, so when that
 *   happens the torrent is re-added selecting only the target file.
 */
export default class RealDebrid extends DebridService {
  static id = 'realdebrid'
  static title = 'Real-Debrid'
  static available = true
  // documented allowance is 250 requests per minute, keep some headroom
  static limits = { reservoir: 200, reservoirRefreshAmount: 200, reservoirRefreshInterval: 60_000, maxConcurrent: 4, minTime: 150 }
  // no cache endpoint any more, so availability has to be probed a hash at a time
  static availabilityCheck = 'probe'

  /**
   * @type {number} Status reads a probe gives a magnet still converting before giving up.
   *
   * Real-Debrid holds the metadata for anything cached, so a cached release reaches file selection
   * within a read or two; one still converting is telling us it is not. Counted in reads rather
   * than seconds so a slow link does not change how many chances it gets. Giving up leaves the
   * release unknown, so the next sweep asks again.
   */
  static probeConversionReads = 2

  /** @param {any} error */
  throttled (error) {
    return super.throttled(error) || throttleCodes.includes(error?.code)
  }

  mapError (status, json) {
    const code = json?.error_code
    const message = errorMessages[code] || json?.error || `Request failed with status ${status}`
    // a blocked or unavailable file also answers 403, and must stay a plain error: an auth error
    // aborts the whole resolve, where one bad file in a pack should only be skipped
    if (authCodes.includes(code) || ((status === 401 || status === 403) && code == null)) return new DebridAuthError(message, { status, code })
    return new DebridError(message, { status, code })
  }

  async validate () {
    const user = await this.request(`${API}/user`)
    if (user?.type !== 'premium') throw new DebridAuthError('Real-Debrid premium is required to stream torrents')
    return { username: user.username, expires: user.expiration }
  }

  /** The whole account in one request, shared by the base class between badges and playback. */
  async fetchListing () {
    return (await this.request(`${API}/torrents?limit=${LIST_LIMIT}`)) || []
  }

  async listAvailability () {
    const torrents = await this.listing()
    const known = new Map()
    for (const torrent of torrents) {
      const state = statusAvailability[torrent.status]
      if (state && torrent.hash) known.set(torrent.hash.toLowerCase(), state)
    }
    return known
  }

  /**
   * The only way Real-Debrid can still be asked about a release: add the magnet and read the
   * status it settles on. A cached torrent reports 'downloaded' within a second of file selection.
   * Costs about five requests, hence the cap in the base class.
   *
   * The torrent is always removed again, and adding a hash the account already holds creates a
   * separate entry, so this can never delete the user's own download.
   * @param {string} hash
   * @returns {Promise<string>}
   */
  async probeAvailability (hash) {
    const magnetURI = RealDebrid.toMagnet(hash)
    if (!magnetURI) throw new DebridError('Not a usable info hash') // no answer, rather than a made up one
    let torrentId = null
    try {
      torrentId = await this.#addAndSelect(magnetURI, () => true, { reads: RealDebrid.probeConversionReads })
      await this.#awaitStatus(torrentId, 'downloaded', this.budget('probe'))
      return Availability.CACHED
    } finally {
      // awaited, so the account is as we found it by the time this answers. Playback and teardown
      // both wait on it, so neither can trip over a half-finished probe
      if (torrentId) await this.release(`${API}/torrents/delete/${torrentId}`)
    }
  }

  async resolve (magnet, { fileFilter = () => true, pickFile, maxFiles = this.config.maxFiles } = {}) {
    const hash = RealDebrid.parseHash(magnet)
    const magnetURI = RealDebrid.toMagnet(magnet)
    let torrentId = null
    let added = false
    try {
      // a cache probe of this same release is about to delete its own torrent, so let it finish
      // first rather than reusing an id that is seconds from disappearing
      await this.probes.get(hash)?.catch(() => {})
      // reuse a torrent that is already on the account instead of adding a duplicate
      const existing = await this.#existingTorrent(hash)
      let info = null
      if (existing?.status === 'waiting_files_selection') {
        // a stale add that never got its files selected, finish the job
        const ids = existing.files.filter(file => fileFilter(file.path)).map(file => file.id)
        await this.request(`${API}/torrents/selectFiles/${existing.id}`, { method: 'POST', body: { files: ids.length ? ids.join(',') : 'all' } })
        torrentId = existing.id
      // mid-conversion counts as not cached here: playback cannot wait for it either way
      } else if (existing && existing.status !== 'downloaded') throw unstreamable(existing.status) ?? new DebridNotCachedError()
      else if (existing) {
        torrentId = existing.id
        info = existing // already confirmed downloaded, with its files and links
      } else {
        torrentId = await this.#addAndSelect(magnetURI, fileFilter)
        added = true
      }
      info ??= await this.#awaitStatus(torrentId, 'downloaded', this.budget('ready'))

      // work out which file playback is after before unrestricting, so a capped pack never drops
      // the wanted episode and archives can be recovered from
      const wanted = info.files.filter(file => fileFilter(file.path)).map(file => ({ id: file.id, path: file.path, size: file.bytes }))
      const target = wanted.length ? (pickFile ? await pickFile(wanted) : [...wanted].sort((a, b) => b.size - a.size)[0]) : null
      let files = await this.#unrestrictLinks(info, fileFilter, maxFiles, target)
      if (target && !files.some(file => file.name === target.path.split('/').pop())) {
        debug(`Re-adding torrent to select only ${target.path}`)
        const retryId = await this.#addAndSelect(magnetURI, null, { fileId: target.id })
        // not awaited: the player is already open waiting on this, cleanup can catch up
        if (added) this.release(`${API}/torrents/delete/${torrentId}`)
        torrentId = retryId
        added = true
        info = await this.#awaitStatus(torrentId, 'downloaded', this.budget('ready'))
        files = await this.#unrestrictLinks(info, fileFilter, 1, null)
        if (!files.length) throw new DebridError('Real-Debrid only serves this torrent as an archive')
      }
      if (!files.length) throw new DebridError('No playable files in this torrent')
      debug(`Resolved ${files.length} files for ${info.filename}`)
      return { hash: info.hash.toLowerCase(), name: info.filename, files }
    } catch (error) {
      // only clean up torrents this call added, never the user's own downloads
      if (added && torrentId) await this.release(`${API}/torrents/delete/${torrentId}`)
      // playback has waited as long as it can, so a magnet still converting is one it cannot use.
      // A probe leaves the same timeout unanswered instead, since Real-Debrid may just be slow
      throw error instanceof DebridTimeoutError ? new DebridNotCachedError() : error
    }
  }

  /**
   * The account's entry for an info hash, or null. Read back by id because the listing can be a
   * minute stale, so one deleted elsewhere must read as absent rather than fail the resolve.
   * @param {string} hash
   * @returns {Promise<any | null>}
   */
  async #existingTorrent (hash) {
    if (!hash) return null
    const listed = (await this.listing()).find(torrent => torrent.hash?.toLowerCase() === hash)
    if (!listed) return null
    try {
      return await this.request(`${API}/torrents/info/${listed.id}`)
    } catch (error) {
      if (error.status !== 404) throw error
      debug(`Account listing named a torrent that is gone (${listed.id}), adding the magnet instead`)
      this.forgetListing()
      return null
    }
  }

  /**
   * Adds a magnet and selects either the files matching the filter or one specific file.
   * @param {string} magnetURI
   * @param {((name: string) => boolean) | null} fileFilter
   * @param {{ fileId?: number, reads?: number }} [opts] - `fileId` selects one file outright,
   *   `reads` caps how many status reads the magnet gets to convert, for probes.
   * @returns {Promise<string>} The new torrent id.
   */
  async #addAndSelect (magnetURI, fileFilter, { fileId, reads } = {}) {
    const torrentId = (await this.request(`${API}/torrents/addMagnet`, { method: 'POST', body: { magnet: magnetURI } }))?.id
    this.forgetListing() // the account has a torrent the remembered listing does not
    try {
      const info = await this.#awaitStatus(torrentId, 'waiting_files_selection', this.budget('select'), reads)
      if (info.status === 'waiting_files_selection') {
        const ids = fileId ? [fileId] : info.files.filter(file => fileFilter(file.path)).map(file => file.id)
        await this.request(`${API}/torrents/selectFiles/${torrentId}`, { method: 'POST', body: { files: ids.length ? ids.join(',') : 'all' } })
      }
      return torrentId
    } catch (error) {
      // awaited, so this call has undone itself by the time it reports failure
      if (torrentId) await this.release(`${API}/torrents/delete/${torrentId}`)
      throw error
    }
  }

  /**
   * Unrestricts a torrent's links into direct stream files. The cached copy may serve fewer links
   * than files selected, so filter by path when the lists align and by filename otherwise.
   * Archives are dropped; the caller recovers via single file selection.
   * @param {any} info
   * @param {(name: string) => boolean} fileFilter
   * @param {number} maxFiles
   * @param {{ path: string } | null} [target] - The file playback wants, kept inside the cap.
   */
  async #unrestrictLinks (info, fileFilter, maxFiles, target) {
    if (!info.links?.length) throw new DebridError('Real-Debrid returned no links for this torrent')
    const selected = info.files.filter(file => file.selected)
    const aligned = info.links.length === selected.length
    const candidates = RealDebrid.windowFiles(aligned
      ? selected.map((file, index) => ({ link: info.links[index], path: file.path, size: file.bytes })).filter(file => fileFilter(file.path))
      : info.links.map(link => ({ link })), target, maxFiles)
    // dead files are skipped by mapFiles; the caller checks the wanted episode came back
    return this.mapFiles(candidates, async ({ link, path, size }) => {
      const unrestricted = await this.request(`${API}/unrestrict/link`, { method: 'POST', body: { link } })
      const name = path?.split('/').pop() || unrestricted.filename
      if (!path && !fileFilter(name)) return null
      if (archiveRx.test(name) && !selected.some(file => file.path.endsWith(name))) return null // RD packed the selection into an archive
      return { name, path: path || `/${name}`, size: unrestricted.filesize || size, url: unrestricted.download, type: unrestricted.mimeType }
    }, candidate => candidate.path || candidate.link)
  }

  /**
   * Polls torrent info until it reaches the wanted status.
   * @param {string} id
   * @param {string} wanted
   * @param {number} timeout - Deadline in milliseconds.
   * @param {number} [reads] - Status reads allowed, for callers that want a fixed number of chances.
   */
  async #awaitStatus (id, wanted, timeout, reads = Infinity) {
    const started = Date.now()
    for (let read = 1; ; read++) {
      const info = await this.request(`${API}/torrents/info/${id}`)
      if (info.status === wanted || (wanted === 'waiting_files_selection' && info.status === 'downloaded')) return info
      const settled = unstreamable(info.status)
      if (settled) throw settled
      // a rare release can sit in magnet_conversion a while, so running out of time proves nothing
      // about it. Callers decide: playback treats it as uncached, a probe leaves it re-checkable
      if (read >= reads || Date.now() - started > timeout) throw new DebridTimeoutError(`Timed out waiting for Real-Debrid (${info.status})`)
      await new Promise(resolve => setTimeout(resolve, this.config.timeouts.poll).unref?.())
    }
  }
}

/**
 * The typed answer a settled status stands for, or null while the torrent is still deciding.
 * @param {string} status
 * @returns {import('./service.js').DebridUnstreamableError | null}
 */
function unstreamable (status) {
  switch (statusAvailability[status]) {
    case Availability.UNAVAILABLE: return new DebridUnavailableError(`Real-Debrid could not process this torrent (${status})`)
    case Availability.AVAILABLE: return new DebridNotCachedError()
    default: return null
  }
}

File: app/lib/infrastructure/debrid/providers/realdebrid.dart
import '../../../domain/models/availability.dart';
import '../../../domain/ports/debrid_client.dart';
import '../../network/transport.dart';
import '../client.dart';
import '../errors.dart';
import '../hash.dart';
import '../window.dart';
import 'json.dart';
import 'provider.dart';

const _api = 'https://api.real-debrid.com/rest/1.0';
const _listLimit = 1000;
const _probeConversionReads = 2;

const _errorMessages = <int, String>{
  8: 'Invalid Real-Debrid API key',
  9: 'Real-Debrid denied the request, check the account permissions',
  21: 'Too many active Real-Debrid downloads, wait for one to finish',
  23: 'This Real-Debrid account has exhausted its traffic',
  34: 'Real-Debrid is rate limiting this account, try again shortly',
  35: 'Real-Debrid will not serve this file, pick a different release',
  36: 'Real-Debrid fair usage limit reached',
};

class _RealDebridDialect extends Dialect {
  const _RealDebridDialect();

  @override
  DebridFailure mapError(int status, Object? json) {
    final numericCode = jsonInt(field(json, 'error_code'));
    final code = numericCode?.toString();
    final message =
        _errorMessages[numericCode] ??
        safeServiceMessage(
          jsonString(field(json, 'error')),
          'Request failed with status $status',
        );
    final auth =
        numericCode == 8 ||
        numericCode == 9 ||
        ((status == 401 || status == 403) && numericCode == null);
    return auth
        ? DebridFailure.auth(message, status: status, code: code)
        : DebridFailure.service(message, status: status, code: code);
  }
}

class _TorrentFile {
  const _TorrentFile({
    required this.id,
    required this.path,
    required this.bytes,
    required this.selected,
  });

  final int id;
  final String path;
  final int bytes;
  final bool selected;
}

class _TorrentInfo {
  const _TorrentInfo({
    required this.id,
    required this.hash,
    required this.filename,
    required this.status,
    required this.files,
    required this.links,
  });

  final String id;
  final String hash;
  final String filename;
  final String status;
  final List<_TorrentFile> files;
  final List<String> links;
}

class _Wanted {
  const _Wanted({required this.id, required this.path, required this.size});

  final int id;
  final String path;
  final int size;
}

class _Candidate {
  const _Candidate({required this.link, this.path, required this.size});

  final String link;
  final String? path;
  final int size;
}

class RealDebridProvider implements DebridProvider {
  RealDebridProvider(String apiKey, HttpTransport transport, Clock clock)
    : client = DebridApiClient(providerConfig, apiKey, transport, clock);

  static const providerConfig = ProviderConfig(
    id: 'realdebrid',
    title: 'Real-Debrid',
    auth: AuthScheme.bearer,
    authParam: 'apikey',
    encoding: BodyEncoding.form,
    nominalLatencyMs: 300,
    maxFiles: 60,
    availabilityCheck: AvailabilityCheck.probe,
    checkAddsMagnets: true,
    maxBatch: 100,
    maxProbes: 10,
    maxConcurrent: 4,
    minTimeMs: 150,
    reservoir: (200, 60000),
  );

  static const _dialect = _RealDebridDialect();

  @override
  final DebridApiClient client;

  @override
  ProviderConfig get config => providerConfig;

  @override
  bool throttled(DebridFailure error) =>
      error.throttled || error.code == '21' || error.code == '34';

  Future<Object?> _get(String url, {CancelToken? cancel}) =>
      client.request(_dialect, url, const RequestOpts(), cancel: cancel);

  Future<Object?> _post(
    String url,
    List<(String, Object?)> body, {
    CancelToken? cancel,
  }) => client.request(
    _dialect,
    url,
    RequestOpts(method: HttpMethod.post, body: body),
    cancel: cancel,
  );

  static (String, RequestOpts) _deleteRequest(String id) => (
    '$_api/torrents/delete/$id',
    const RequestOpts(method: HttpMethod.delete),
  );

  Future<void> _release(String id) async {
    final (url, opts) = _deleteRequest(id);
    await client.release(_dialect, url, opts);
  }

  Future<Object?> _fetchListing({CancelToken? cancel}) => client.listing(
    false,
    () => _get('$_api/torrents?limit=$_listLimit', cancel: cancel),
  );

  Future<_TorrentInfo?> _existingTorrent(
    String? hash, {
    CancelToken? cancel,
  }) async {
    if (hash == null) return null;
    final listing = await _fetchListing(cancel: cancel);
    String? id;
    for (final torrent in jsonList(listing)) {
      if ((jsonString(field(torrent, 'hash')) ?? '').toLowerCase() == hash) {
        id = jsonString(field(torrent, 'id'));
        break;
      }
    }
    if (id == null) return null;
    try {
      return _parseInfo(await _get('$_api/torrents/info/$id', cancel: cancel));
    } on DebridFailure catch (error) {
      if (error.status == 404) {
        client.forgetListing();
        return null;
      }
      rethrow;
    }
  }

  Future<void> _selectFiles(
    String id,
    List<int> ids, {
    CancelToken? cancel,
  }) async {
    final files = ids.isEmpty ? 'all' : ids.join(',');
    await _post('$_api/torrents/selectFiles/$id', [
      ('files', files),
    ], cancel: cancel);
  }

  Future<String> _addAndSelect(
    String magnet, {
    FileFilter? fileFilter,
    int? fileId,
    int? reads,
    CancelToken? cancel,
  }) async {
    final added = await _post('$_api/torrents/addMagnet', [
      ('magnet', magnet),
    ], cancel: cancel);
    final id = nonEmptyString(field(added, 'id'));
    if (id == null) {
      throw const DebridFailure.service('Real-Debrid returned no torrent id');
    }
    client.forgetListing();
    final guard = OrphanGuard(client);
    final (deleteUrl, deleteOpts) = _deleteRequest(id);
    guard.arm(deleteUrl, deleteOpts);
    try {
      final info = await _awaitStatus(
        id,
        'waiting_files_selection',
        client.budget(config.timeouts.selectMs),
        reads: reads,
        cancel: cancel,
      );
      if (info.status == 'waiting_files_selection') {
        final ids = fileId == null
            ? [
                for (final file in info.files)
                  if (fileFilter == null || fileFilter(file.path)) file.id,
              ]
            : [fileId];
        await _selectFiles(id, ids, cancel: cancel);
      }
      guard.disarm();
      return id;
    } on CancelledException {
      guard.settle();
      rethrow;
    } catch (_) {
      guard.disarm();
      await _release(id);
      rethrow;
    }
  }

  Future<_TorrentInfo> _awaitStatus(
    String id,
    String wanted,
    int timeoutMs, {
    int? reads,
    CancelToken? cancel,
  }) async {
    final started = client.clock.nowMs;
    var read = 0;
    while (true) {
      read += 1;
      final info = _parseInfo(
        await _get('$_api/torrents/info/$id', cancel: cancel),
      );
      if (info.status == wanted ||
          (wanted == 'waiting_files_selection' &&
              info.status == 'downloaded')) {
        return info;
      }
      final settled = _unstreamable(info.status);
      if (settled != null) throw settled;
      if ((reads != null && read >= reads) ||
          client.clock.nowMs - started > timeoutMs) {
        throw DebridFailure.timeout(
          'Timed out waiting for Real-Debrid (${info.status})',
        );
      }
      await raced(client.clock.sleepMs(config.timeouts.pollMs), cancel);
    }
  }

  Future<List<DebridFileInfo>> _unrestrictLinks(
    _TorrentInfo info,
    FileFilter filter,
    int maxFiles,
    _Wanted? target, {
    CancelToken? cancel,
  }) async {
    if (info.links.isEmpty) {
      throw const DebridFailure.service(
        'Real-Debrid returned no links for this torrent',
      );
    }
    final selected = [
      for (final file in info.files)
        if (file.selected) file,
    ];
    final aligned = info.links.length == selected.length;
    final candidates = <_Candidate>[];
    if (aligned) {
      for (var index = 0; index < selected.length; index++) {
        final file = selected[index];
        if (filter(file.path)) {
          candidates.add(
            _Candidate(
              link: info.links[index],
              path: file.path,
              size: file.bytes,
            ),
          );
        }
      }
    } else {
      candidates.addAll([
        for (final link in info.links) _Candidate(link: link, size: 0),
      ]);
    }
    int? targetIndex;
    if (target != null) {
      final found = candidates.indexWhere(
        (candidate) => candidate.path == target.path,
      );
      if (found >= 0) targetIndex = found;
    }
    return mapFiles(windowFiles(candidates, targetIndex, maxFiles), (
      candidate,
    ) async {
      final unrestricted = await _post('$_api/unrestrict/link', [
        ('link', candidate.link),
      ], cancel: cancel);
      final name = candidate.path == null
          ? nonEmptyString(field(unrestricted, 'filename'))
          : basename(candidate.path!);
      if (name == null || name.isEmpty) return null;
      if (candidate.path == null && !filter(name)) return null;
      if (_isArchive(name) &&
          !selected.any((file) => file.path.endsWith(name))) {
        return null;
      }
      final unrestrictedSize = jsonInt(field(unrestricted, 'filesize')) ?? 0;
      return DebridFileInfo(
        name: name,
        path: candidate.path ?? '/$name',
        size: unrestrictedSize > 0 ? unrestrictedSize : candidate.size,
        url: jsonString(field(unrestricted, 'download')) ?? '',
        type: jsonString(field(unrestricted, 'mimeType')),
      );
    });
  }

  @override
  Future<AccountInfo> validate({CancelToken? cancel}) async {
    final user = await _get('$_api/user', cancel: cancel);
    if (jsonString(field(user, 'type')) != 'premium') {
      throw const DebridFailure.auth(
        'Real-Debrid premium is required to stream torrents',
      );
    }
    return AccountInfo(
      username: jsonString(field(user, 'username')) ?? '',
      expires: jsonString(field(user, 'expiration')),
    );
  }

  @override
  Future<Map<String, Availability>> listAvailability({
    CancelToken? cancel,
  }) async {
    final known = <String, Availability>{};
    for (final torrent in jsonList(await _fetchListing(cancel: cancel))) {
      final status = jsonString(field(torrent, 'status'));
      final state = status == null ? null : _statusAvailability(status);
      final hash = nonEmptyString(field(torrent, 'hash'))?.toLowerCase();
      if (state != null && hash != null) known[hash] = state;
    }
    return known;
  }

  @override
  Future<Map<String, Availability>> checkAvailabilityBatch(
    List<String> hashes, {
    CancelToken? cancel,
  }) => Future.error(
    const DebridFailure.service(
      'Real-Debrid has no cache endpoint; availability is probed per hash',
    ),
  );

  @override
  Future<Availability> probeAvailability(
    String hash, {
    CancelToken? cancel,
  }) async {
    final magnet = toMagnet(hash);
    if (magnet == null) {
      throw const DebridFailure.service('Not a usable info hash');
    }
    final id = await _addAndSelect(
      magnet,
      fileFilter: (_) => true,
      reads: _probeConversionReads,
      cancel: cancel,
    );
    final guard = OrphanGuard(client);
    final (url, opts) = _deleteRequest(id);
    guard.arm(url, opts);
    try {
      final result = await _awaitStatus(
        id,
        'downloaded',
        client.budget(config.timeouts.probeMs),
        cancel: cancel,
      );
      await raced(_release(id), cancel);
      guard.disarm();
      return result.status == 'downloaded'
          ? Availability.cached
          : Availability.unknown;
    } on CancelledException {
      guard.settle();
      rethrow;
    } catch (_) {
      await _release(id);
      guard.disarm();
      rethrow;
    }
  }

  @override
  Future<ResolvedFiles> resolve(
    String magnet,
    ResolveOptions opts, {
    CancelToken? cancel,
  }) async {
    final hash = parseHash(magnet);
    final magnetUri = toMagnet(magnet);
    if (magnetUri == null) {
      throw const DebridFailure.service('Not a usable info hash');
    }
    bool filter(String path) => opts.fileFilter?.call(path) ?? true;
    final maxFiles = opts.maxFiles ?? config.maxFiles;
    String? added;
    final guard = OrphanGuard(client);
    try {
      final existing = await _existingTorrent(hash, cancel: cancel);
      String id;
      _TorrentInfo? info;
      if (existing?.status == 'waiting_files_selection') {
        await _selectFiles(existing!.id, [
          for (final file in existing.files)
            if (filter(file.path)) file.id,
        ], cancel: cancel);
        id = existing.id;
      } else if (existing != null && existing.status != 'downloaded') {
        throw _unstreamable(existing.status) ?? const DebridFailure.notCached();
      } else if (existing != null) {
        id = existing.id;
        info = existing;
      } else {
        id = await _addAndSelect(magnetUri, fileFilter: filter, cancel: cancel);
        added = id;
        final (url, deleteOpts) = _deleteRequest(id);
        guard.arm(url, deleteOpts);
      }
      info ??= await _awaitStatus(
        id,
        'downloaded',
        client.budget(config.timeouts.readyMs),
        cancel: cancel,
      );

      final wanted = [
        for (final file in info.files)
          if (filter(file.path))
            _Wanted(id: file.id, path: file.path, size: file.bytes),
      ];
      _Wanted? target;
      if (wanted.isNotEmpty) {
        if (opts.pickFile != null) {
          final index = opts.pickFile!([
            for (final file in wanted) (file.id, file.path, file.size),
          ]);
          if (index != null && index < wanted.length) target = wanted[index];
        } else {
          target = wanted.reduce(
            (best, file) => file.size > best.size ? file : best,
          );
        }
      }

      var files = await _unrestrictLinks(
        info,
        filter,
        maxFiles,
        target,
        cancel: cancel,
      );
      if (target != null &&
          !files.any((file) => file.name == basename(target!.path))) {
        final retryId = await _addAndSelect(
          magnetUri,
          fileId: target.id,
          cancel: cancel,
        );
        if (added != null) await _release(id);
        id = retryId;
        added = retryId;
        final (url, deleteOpts) = _deleteRequest(retryId);
        guard.arm(url, deleteOpts);
        info = await _awaitStatus(
          retryId,
          'downloaded',
          client.budget(config.timeouts.readyMs),
          cancel: cancel,
        );
        files = await _unrestrictLinks(info, filter, 1, null, cancel: cancel);
        if (files.isEmpty) {
          throw const DebridFailure.service(
            'Real-Debrid only serves this torrent as an archive',
          );
        }
      }
      if (files.isEmpty) {
        throw const DebridFailure.service('No playable files in this torrent');
      }
      final resolved = ResolvedFiles(
        hash: info.hash,
        name: info.filename,
        files: secureFiles(files, config.title),
        targetPath: target?.path,
      );
      guard.disarm();
      return resolved;
    } on CancelledException {
      guard.settle();
      rethrow;
    } on DebridFailure catch (error) {
      guard.disarm();
      if (added != null) await _release(added);
      if (error.kind == DebridErrorKind.timeout) {
        throw const DebridFailure.notCached();
      }
      rethrow;
    } catch (_) {
      guard.disarm();
      if (added != null) await _release(added);
      rethrow;
    }
  }

  @override
  Future<void> retryCleanup() => client.retryCleanup(_dialect);
}

Availability? _statusAvailability(String status) => switch (status) {
  'downloaded' => Availability.cached,
  'queued' ||
  'downloading' ||
  'uploading' ||
  'compressing' => Availability.available,
  'magnet_error' || 'error' || 'virus' || 'dead' => Availability.unavailable,
  _ => null,
};

DebridFailure? _unstreamable(String status) =>
    switch (_statusAvailability(status)) {
      Availability.unavailable => DebridFailure.unavailable(
        'Real-Debrid could not process this torrent ($status)',
      ),
      Availability.available => const DebridFailure.notCached(),
      _ => null,
    };

_TorrentInfo _parseInfo(Object? value) => _TorrentInfo(
  id: jsonString(field(value, 'id')) ?? '',
  hash: (jsonString(field(value, 'hash')) ?? '').toLowerCase(),
  filename: jsonString(field(value, 'filename')) ?? '',
  status: jsonString(field(value, 'status')) ?? '',
  files: [
    for (final file in jsonList(field(value, 'files')))
      _TorrentFile(
        id: jsonInt(field(file, 'id')) ?? 0,
        path: jsonString(field(file, 'path')) ?? '',
        bytes: jsonInt(field(file, 'bytes')) ?? 0,
        selected: jsonTruthy(field(file, 'selected')),
      ),
  ],
  links: [
    for (final link in jsonList(field(value, 'links')))
      if (link is String) link,
  ],
);

bool _isArchive(String name) {
  final lower = name.toLowerCase();
  return lower.endsWith('.rar') ||
      lower.endsWith('.zip') ||
      lower.endsWith('.7z');
}
